### PR TITLE
Computed options

### PR DESCRIPTION
**Pull request by xcambar**
A follow up on #230

### CODE REVIEW

**Comment by offirgolan** [comment]
@xcambar thank you! 

As a follow up to you previous question, I've decided to completely remove `options as functions` and replace them with CPs. This should be a major API change that will bring this addon to `3.0`.

File: addon/validations/factory.js [hunk 1]
```diff
@@ -430,7 +430,7 @@ function createTopLevelPropsMixin(validatableAttrs) {
 function getCPDependentKeysFor(attribute, validations, owner) {
   let dependentKeys = validations.map(validation => {
     const type = validation._type;
-    const options = validation.options;
+    const options = getWithDefault(validation, 'options', Ember.Object.create());
```
Comment (offirgolan) [inline addon/validations/factory.js]: `validation.options` is valid here. This should be an will always be a pojo. Please revert this change back.

File: addon/validations/factory.js [hunk 2]
```diff
@@ -445,8 +445,8 @@ function getCPDependentKeysFor(attribute, validations, owner) {
     const cpDependents = Object.keys(options).reduce((arr, key) => {
       let option = options[key];
 
-      if(typeof option === 'object' && get(option, 'isDescriptor')) {
-        return arr.concat(get(option, '_dependentKeys').map(d => {
+      if(get(option, 'isDescriptor')) {
```
Comment (offirgolan) [inline addon/validations/factory.js]: Please revert this changes back as well. The checks I made are intentional.

File: addon/validations/factory.js [hunk 3]
```diff
@@ -445,8 +445,8 @@ function getCPDependentKeysFor(attribute, validations, owner) {
...
+        return arr.concat(getWithDefault(option, '_dependentKeys', []).map(d => {
```
Comment (offirgolan) [inline addon/validations/factory.js]: No need for getWithDefault here. Something like

`(get(option, '_dependentKeys') || []).map` is fine since this is not an ember object

File: addon/validations/result-collection.js [hunk 4]
```diff
@@ -271,7 +270,7 @@ export default Ember.Object.extend({
    * @readOnly
    * @type {Ember.ComputedProperty | Object}
    */
-  options: computed('_contentValidators.[]', '_contentValidators.@each._cachedOptions', function () {
+  options: computed('_contentValidators.[]', '_contentValidators.[]', function () {
```
Comment (offirgolan) [inline addon/validations/result-collection.js]: For the 2nd dependent: `'_contentValidators.@each.options'`

File: addon/validations/result-collection.js [hunk 5]
```diff
@@ -322,18 +321,18 @@ export default Ember.Object.extend({
...
-          options[type] = [options[type], vOpts];
+          options.set(type, [options.get(type), vOpts]);
         }
       } else {
-        options[type] = vOpts;
+        options.set(type, vOpts);
       }
       return options;
-    }, {});
+    }, Ember.Object.create());
```
Comment (offirgolan) [inline addon/validations/result-collection.js]: Id rather not have this be an Ember.Object. pojo is fine here.

File: addon/validators/base.js [hunk 6]
```diff
@@ -116,10 +109,10 @@ const Base = Ember.Object.extend({
 
     // Overwrite the validator's value method if it exists in the options and remove it since
     // there is no need for it to be passed around
-    this.value = builtOptions.value || this.value;
+    this.value = getWithDefault(builtOptions, 'value', get(this, 'value'));
     delete builtOptions.value;
 
-    const OptionsClass = Ember.Object.extend(builtOptions, {
+    const OptionsClass = Ember.Object.extend({
```
Comment (offirgolan) [inline addon/validators/base.js]: Please revert all changes made to `OptionsClass`.

File: addon/validators/base.js [hunk 7]
```diff
@@ -116,10 +109,10 @@ const Base = Ember.Object.extend({
 
     // Overwrite the validator's value method if it exists in the options and remove it since
     // there is no need for it to be passed around
-    this.value = builtOptions.value || this.value;
+    this.value = getWithDefault(builtOptions, 'value', get(this, 'value'));
```
Comment (offirgolan) [inline addon/validators/base.js]: No need to use getWithDefault here. builtOptions it just a pojo

File: addon/validators/base.js [hunk 8]
```diff
@@ -128,14 +121,19 @@ const Base = Ember.Object.extend({
...
+        opts.setProperties(appliedOpts);
+        return opts;
       }
     });
 
-    return Object.freeze(OptionsClass.create());
+    var optionsInstance = OptionsClass.create();
+    optionsInstance.setProperties(builtOptions);
+    return optionsInstance;
   },
```
Comment (offirgolan) [inline addon/validators/base.js]: @xcambar please revert all changes made to `base.js`. Everything I did in there was well thought out and intentional. If you would like, I can give you a detailed explanation as to why I'm doing all of this. If you're on ember slack, we can just chat on there 😸

File: addon/validators/length.js [hunk 9]
```diff
@@ -55,25 +55,30 @@ export default Base.extend({
   },
 
   validate(value, options) {
+    if (isEmpty(Object.keys(options))) {
```
Comment (offirgolan) [inline addon/validators/length.js]: No need for this block.

**Comment by offirgolan** [comment]
@xcambar sorry for bombarding your PR. Im a bit picky. If you have any questions, feel free to ask me on here or on ember slack.

**Comment by offirgolan** [comment]
@xcambar i've been working on this all morning 😛 . Im farther ahead then this PR so im gonna go ahead and close this! Feel free to review the branch and reach out if you notice something wonky. 

Thanks for bringing this to my attention! Im super excited for the release of this lol. Ill make sure to add a special thanks to you in the release notes. 😺

**Comment by xcambar** [comment]
I'll look at the branch today and I'll let you know. Thank you for being so receptive to the idea.
I'm as excited as you are and really looking forward to the feature!
